File: HuffmanCoding/CharacterFrequency.cs
using System;

namespace HuffmanCoding
{
    public class CharacterFrequency : IComparable
    {
        private char character;
        private int frequency;

        public char Character
        {
            get
            {
                return character;
            }
            set
            {
                character = value;
            }
        }

        public int Frequency
        {
            get { return frequency; }
            set
            {
                if (value >= 0) frequency = value;
                else throw new ArgumentOutOfRangeException();
            }
        }

        public CharacterFrequency()
        {
            Character = '\0';
            Frequency = 0;
        }

        public CharacterFrequency(char ch)
        {
            Character = ch;
            Frequency = 0;
        }

        public CharacterFrequency(char ch, int count)
        {
            Character = ch;
            Frequency = count;
        }

        public CharacterFrequency(CharacterFrequency cf)
        {
            Character = cf.Character;
            Frequency = cf.Frequency;
        }

        public void Increment()
        {
            frequency++;
        }

        public static CharacterFrequency operator +(CharacterFrequency firstAddend, CharacterFrequency secondAddend)
        {
            CharacterFrequency sum = new CharacterFrequency();

            if (firstAddend.Character == secondAddend.Character)
            {
                sum.Character = firstAddend.Character;
                sum.Frequency = firstAddend.Frequency + secondAddend.Frequency;
            }
            else
            {
                sum.Character = '\0';
                sum.Frequency = firstAddend.Frequency + secondAddend.Frequency;
            }

            return sum;
        }

        public static Boolean operator <(CharacterFrequency firstTerm, CharacterFrequency secondTerm)
        {
            return firstTerm.Frequency < secondTerm.Frequency;
        }

        public static Boolean operator >(CharacterFrequency firstTerm, CharacterFrequency secondTerm)
        {
            return firstTerm.Frequency > secondTerm.Frequency;
        }

        public int CompareTo(object obj)
        {
            if (obj == null) return 1;
            if (obj == this) return 0;

            CharacterFrequency cf = obj as CharacterFrequency;
            if (cf != null)
                return this.frequency.CompareTo(cf.frequency);
            else throw new ArgumentException("Object is not a Character Frequency");
        }

        public override Boolean Equals(object obj)
        {
            if (obj == null) return false;
            if (obj == this) return true;
            if (!(obj.GetType() == GetType())) return false;

            CharacterFrequency cf = obj as CharacterFrequency;

            return Character.Equals(cf.Character);
        }

        public override int GetHashCode()
        {
            return (int)Character;
        }

        public override string ToString()
        {
            return String.Format($"{character}|{(int)character}|{frequency}");
        }
    }
}

File: src/me/victorfrye/huffmancoding/model/CharacterFrequency.java
package me.victorfrye.huffmancoding.model;

import static java.util.Objects.hash;

/**
 * A class to identify a character and count occurrences of the character.
 * <p>
 * This class has two variables: 'character' and 'frequency'.  The character identifies the class while the frequency maintains a count of occurrences.  Constructors are included for default, character only, character + frequency, and a CharacterFrequency object.  Comparison of CharacterFrequency objects is done only by the character variable.
 * </p>
 *
 * @author Victor W. Frye <iam@victorfrye.me>
 */
public class CharacterFrequency {
  private char character;
  private int frequency;

  public CharacterFrequency() {
    this.character = '\0';
    this.frequency = 0;
  }

  public CharacterFrequency(char c) {
    this.character = c;
    this.frequency = 0;
  }

  public CharacterFrequency(char c, int i) {
    this.character = c;
    this.setFrequency(i);
  }

  public CharacterFrequency(CharacterFrequency cf) {
    this.character = cf.getCharacter();
    this.frequency = cf.getFrequency();
  }

  public char getCharacter() {
    return character;
  }

  public void setCharacter(char c) {
    this.character = c;
  }

  public CharacterFrequency character(CharacterFrequency cf) {
    this.character = cf.character;
    return this;
  }

  public int getFrequency() {
    return frequency;
  }

  public void setFrequency(int i) {
    if (i < 0) {
      throw new IllegalArgumentException();
    } else {
      this.frequency = i;
    }
  }

  public CharacterFrequency frequency(CharacterFrequency cf) {
    this.frequency = cf.frequency;
    return this;
  }

  public void increment() {
    frequency++;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (!(o instanceof CharacterFrequency)) return false;
    CharacterFrequency that = (CharacterFrequency) o;
    return character == that.character;
  }

  @Override
  public int hashCode() {
    return hash(character);
  }

  @Override
  public String toString() {
    return String.format("CharacterFrequency{character=%s, frequency=%d}", character, frequency);
  }
}
